display_list/display_list_color_source.h

```diff
@@ -0,0 +1,668 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_DISPLAY_LIST_DISPLAY_LIST_COLOR_SOURCE_H_
+#define FLUTTER_DISPLAY_LIST_DISPLAY_LIST_COLOR_SOURCE_H_
+
+#include "flutter/display_list/display_list.h"
+#include "flutter/display_list/display_list_attributes.h"
+#include "flutter/display_list/types.h"
+#include "flutter/fml/logging.h"
+#include "third_party/skia/include/core/SkShader.h"
+#include "third_party/skia/include/effects/SkGradientShader.h"
+
+namespace flutter {
+
+class DlColorColorSource;
+class DlImageColorSource;
+class DlLinearGradientColorSource;
+class DlRadialGradientColorSource;
+class DlConicalGradientColorSource;
+class DlSweepGradientColorSource;
+class DlUnknownColorSource;
+
+enum class DlColorSourceType {
+  kColor,
+  kImage,
+  kLinearGradient,
+  kRadialGradient,
+  kConicalGradient,
+  kSweepGradient,
+  kUnknown
+};
+
+enum class DlTileMode {
+  // Replicate the edge color if the shader draws outside of its
+  // original bounds.
+  kClamp,
+
+  // Repeat the shader's image horizontally and vertically.
+  kRepeat,
+
+  // Repeat the shader's image horizontally and vertically, alternating
+  // mirror images so that adjacent images always seam.
+  kMirror,
+
+  // Only draw within the original domain, return transparent-black everywhere
+  // else.
+  kDecal,
+};
+
+class DlColorSource
```

Comment:
Can you add a headerdoc section to this. Per my understanding:

```
Objects that define how pixels within the coverage area of rendering operations are generated or transformed. The equivalent concept in Skia is a an SkShader. 
```

But you have more context right now about why this name was chosen and how this maps to Skia concepts. So it would be good to have that piece of documentation in there.

Reply:
Done. I also moved DlTileMode to its own header.

Reply:
```
// The DisplayList ColorSource class. This class implements all of the
// facilities and adheres to the design goals of the |DlAttribute| base
// class.
// 
// The role of the DlColorSource is to provide color information for
// the pixels of a rendering operation. The object is essentially the
// origin of all color being rendered, though its output may be
// modified or transformed by geometric coverage data, the filter
// attributes, and the final blend with the pixels in the destination.
```